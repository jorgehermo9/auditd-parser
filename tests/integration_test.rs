use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
