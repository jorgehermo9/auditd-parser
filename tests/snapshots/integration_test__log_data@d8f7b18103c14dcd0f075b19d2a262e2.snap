---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/go-libaudit/audit-rhel7.log:7"
expression: result
info: "type=CRED_REFR msg=audit(1489640461.198:480): pid=1402 uid=0 auid=0 ses=2 subj=system_u:system_r:crond_t:s0-s0:c0.c1023 msg='op=PAM:setcred grantors=pam_env,pam_unix acct=\"root\" exe=\"/usr/sbin/crond\" hostname=? addr=? terminal=cron res=success'"
---
{
  "record_type": "CRED_REFR",
  "timestamp": 1489640461198,
  "id": 480,
  "fields": {
    "auid": "root",
    "msg": {
      "acct": "root",
      "addr": "?",
      "exe": "/usr/sbin/crond",
      "grantors": [
        "pam_env",
        "pam_unix"
      ],
      "hostname": "?",
      "op": "PAM:setcred",
      "res": "success",
      "terminal": "cron"
    },
    "pid": "1402",
    "ses": "2",
    "subj": {
      "level": {
        "category": "c0.c1023",
        "sensitivity": "s0-s0"
      },
      "module": "SELinux",
      "role": "system_r",
      "type": "crond_t",
      "user": "system_u"
    },
    "uid": "root"
  }
}
